<h2> Комментарий ревьюера <a class="tocSkip"></h2>

Дмитрий, привет!

Меня зовут Олег Михеев, и я буду проверять твой проект. Предлагаю общаться на "ты", но если тебе будет некомфортно, обязательно сообщи, и мы перейдем на "вы". Моя главная цель - не указать на твои ошибки, а поделиться опытом и подготовить тебя к работе аналитиком. Увидев у тебя ошибку, я постараюсь указать на её наличие и дам возможность найти и самостоятельно исправить её.

В проверке проекта я буду использовать разные цвета для обозначения комментариев. Например:

<br/>

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Так выделены удачные решения, на которые можно опираться в будущем.
</div>

<br/>

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Когда решение на отдельном шаге станет лучше, если внести небольшие коррективы.
</div>


<br/>
<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

 А так выделены решения, которые необходимо доработать, чтобы я мог принять проект.
</div>
    
  
Если вносишь изменения в проект, указывай, пожалуйста, это в своих комментариях. Будет удобно, если обозначишь их заметным цветом. Например, вот так:
       
<div class="alert alert-block alert-info">
    
<h3> Комментарий студента  <a class="tocSkip"></h3>
    
Можешь кликнуть сюда два раза и скопировать код комментария. Также если что-то непонятно или остались вопросы по ревью, обязательно пиши.
    
</div>
    
Пожалуйста, не удаляй и не перемещай мои комментарии, чтобы не произошло путаницы, и я мог быстрее проверить проект.
<br/>

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью**

Здорово, что удалось пройти по всем этапам исследования. Ты уже неплохо освоил теорию, но некоторые моменты необходимо доработать, чтобы можно было идти дальше.

**Что нужно сделать, чтобы я мог принять проект**:

* Вывести сводные таблицы с количеством клиентов, количеством должников и процентом должников для всех вопросов исследовательской части.
* Пересмотреть выводы там, где необходимо, с учётом объёма категорий клиентов
* Уточнить ответ в вопросе о медиане
* Написать подробный финальный вывод, учитывающий все группы клиентов с процентными показателями и информацию о предобработке

**Успехов! Жду проект на повторное ревью!**

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью v2**

Спасибо, что постарался учесть все замечания и рекомендации.

Осталось совсем немного. Чтобы я мог принять проект, опиши финальный вывод более подробно, пожалуйста.

**Буду ждать твой проект на следующую итерацию!**

<div style="border:solid Chocolate 2px; padding: 40px">

**Общий вывод по ревью v3**

Все критические замечания учтены!

Придерживайся тех же принципов работы в следующих проектах, и это облегчит тебе взаимодействие с потенциальными коллегами и позволит быстрее сдавать учебные проекты.

**Проект принимаю. Успехов в обучении!**

# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
import seaborn as sb
import matplotlib.pyplot as plt
try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [32]:
# Категоризируем данные по наличию детей в семье
# Cоздаем функцию children_func 
def children_func(children):
    if (children == 1) | (children == 2):
        return 'есть дети'
    if children >= 3:
        return 'многодетная семья'
    return 'нет детей'
data['children_category'] = data['children'].apply(children_func)
# Сгруппируем данные по новой категории и найдем отношение наличия детей и задолжности
#data_grouped_debt_ratio = data.groupby('children_category')['debt'].sum() / data.groupby('children_category')['debt'].count()
# Преобразуем данные в процентый формат, для наглядности
#data_grouped_debt_ratio = data_grouped_debt_ratio.map('{:.1%}'.format)
# Создаем таблицу
children_debt_ratio = data.pivot_table(index='children_category', values='debt', aggfunc=['sum','count','mean'])
# Добовляем столбец
children_debt_ratio



sum  count      mean
                   debt   debt      debt
children_category                       
есть дети           638   6860  0.093003
многодетная семья    31    380  0.081579
нет детей          1063  14091  0.075438

**Вывод: Наличие детей имеет зависимость с наличием задолжностей в семье.** 
Хотя и можно заметить что многодетные семьи(семьи, где детей 3 или больше) чаще платят в срок, чем семьи с 1 или 2 детьми, возможно это связано с социальными выплатами, программами поддержки, или колличество клиентов в категории "многодетная семья" недостаточно для общего вывода.

<div class="alert alert-warning">
    <h3> Комментарий ревьюера v2 ⚠️<a class="tocSkip"> </h3>

Да, всё так, правильно, что обращаешь внимание на размеры категорий клиентов.
        
В aggfunc можно передавать список функций. Например, aggfunc=['func1', 'func2', 'func3']

</div>

<div class="alert alert-danger">
    <h3> Комментарий ревьюера ❌<a class="tocSkip"> </h3>

Если говорить только о проценте просроченных платежей, то вывод верный. Но его нельзя назвать объективным, так как нет информации о объёме выборок категорий заёмщиков. Некоторые группы могут не обладать достаточным объёмом для объективных выводов. Это нужно проверить

И я хочу убедиться, что тебе удалось освоить сводные таблицы на хорошем уровне. Потому что в дальнейшем ты будешь пользоваться ими постоянно. Выведи, пожалуйста, общую сводную таблицу с количеством клиентов, количеством должников и процентом должников. В параметрах сводных таблиц укажи сразу несколько агрегирующих функций, пожалуйста.

Даю несколько ссылок, в которых есть описание, как это можно сделать:       
* С помощью pivot_table и аргумента aggfunc https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.pivot_table.html
* С помощью groupby и функции .agg() https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html под заголовком The aggregate() method

При написании выводов учитывай, пожалуйста, репрезентативность групп.

Добавь такие таблицы для всех вопросов исследовательской части.
        
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Воспользуемся решением из предыдушего задания.
data_status_debt_ratio = data.groupby('family_status')['debt'].sum() / data.groupby('family_status')['debt'].count()
# Преобразуем данные в процентый формат, для наглядности
data_status_debt_ratio = data_status_debt_ratio.map('{:.1%}'.format)
# Создаем таблицу
family_debt_ratio = data.pivot_table(index='family_status', columns='debt', values='gender', aggfunc='count')
# Обозначим столбцы (0 и 1, будут не очень понятны)
family_debt_ratio.columns=['no_debt', 'debt']
# Добовляем столбец 
family_debt_ratio['ratio'] = data_status_debt_ratio
family_debt_ratio.sort_values(by='ratio',ascending=False)



**Вывод: Исходя из сгруппированных данных, мы можем сделать вывод, что человек, не состоящий в отношениях, чаще всего имеет задолжность по кредиту.** Возможно, у этой категории заемщиков, более свободное распоряжение средствами и нет общих накоплений. Люди в браке, имея, возможно, общий доход, чаще платят в срок, как и люди в разводе. Да, зависимость есть, выборка довольно репрезентативная.

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

 С выводом по проценту согласен. Но добавь, пожалуйста, сводную таблицу и сделай оценку репрезентативности групп. Измени, пожалуйста, вывод, если потребуется.
</div>

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 👍
</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
# Ваш код будет здесь. Вы можете создавать новые ячейки.
# Вспомним категоризацию клиентов из предыдущих заданий.
#0–30000 — 'E';
#30001–50000 — 'D';
#50001–200000 — 'C';
#200001–1000000 — 'B';
#1000001 и выше — 'A'.
df_total_income_debt = data.groupby('total_income_category')['debt'].sum() / data.groupby('total_income_category')['debt'].count()
df_total_income_debt = df_total_income_debt.map('{:.1%}'.format)
# Создаем таблицу
total_income_debt_ratio = data.pivot_table(index='total_income_category', columns='debt', values='gender', aggfunc='count')
# Обозначим столбцы (0 и 1, будут не очень понятны)
total_income_debt_ratio.columns=['no_debt', 'debt']
# Добовляем столбец 
total_income_debt_ratio['ratio'] = df_total_income_debt
# Если честно не понял, как на практике использовать данную функцию и правильно ее применить.
total_income_debt_ratio['representativeness'] = pd.qcut(total_income_debt_ratio['no_debt'], q=3, labels=['c', 'b', 'a'])
total_income_debt_ratio.sort_values(by='ratio',ascending=False)

**Вывод:** Исходя из неоднозначной выборки(очень маленькая репрезентативность у категорий E A D. Но даже по таким данным можно сделать выводы, что наименьшее доверие будут вызывать клиенты с категорией (C), а наибольшее у (B). 

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 Да, верно.
</div>

<div class="alert alert-warning">
    <h3> Комментарий ревьюера ⚠️<a class="tocSkip"> </h3>

Всё так, но обрати внимание, в этой таблице группы крайне несбалансированны. Поэтому стоит учитывать только две самые многочисленные - B и С, так как они включают примерно 98 процентов данных от общей выборки

Можно решить это задание через разделение выборки на равные по объёму части. Для этого подойдёт функция pd.qcut().

Даю ссылку:
    
* https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html

Метод qcut() делит числовой столбец на равные по объёму части на основании квантилей. Квантили — это статистические показатели, которые разделяют набор данных на равные части. Квантили включают квартили (25%, 50% и 75%), децили (каждые 10%) и процентили (каждые 1%). Приведу пример разделения на 3 равных отрезка, находящихся на отметках 33,3% и 66,7%.
        
    pd.qcut(data['column'], q=3, labels=['one', 'two', 'three'])
        
Можно не указывать названия отрезков с помощью labels. Тогда названия отрезкам присвоятся в численном виде.
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
# Ваш код будет здесь. Вы можете создавать новые ячейки
df_debt_purpose = data.groupby('purpose_category')['debt'].sum() / data.groupby('purpose_category')['debt'].count()
df_debt_purpose = df_debt_purpose.map('{:.1%}'.format)
# Создаем таблицу
purpose_debt_ratio = data.pivot_table(index='purpose_category', columns='debt', values='gender', aggfunc='count')
# Обозначим столбцы (0 и 1, будут не очень понятны)
purpose_debt_ratio.columns=['no_debt', 'debt']
# Добовляем столбец 
purpose_debt_ratio['ratio'] = df_debt_purpose
purpose_debt_ratio.sort_values(by='ratio',ascending=False)

**Вывод:** Выборка хорошая, репрезентативность по категорям нигде не завышена, значит можно сделать качественный вывод. Кредиты с целью покупки недвижимости или проведение свадьбы чаще платят в срок, нежели покупки автомобилей или образования.

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 Верно. А почему кредиты на недвижимость и свадьбу выплачиваются лучше, как ты считаешь?
</div>

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

 Вывод только по процентам верный. Но также нужно оценить репрезентативность групп после добавления сводной таблицы с параметрами из первого комментария.
</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Существуют 2 проблемы. Техническая ошибка(повреждение файла) или Человеческий фактор(неверный ввод данных)

<div class="alert alert-success">
<h3> Комментарий ревьюера ✔️<a class="tocSkip"> </h3>

Абсолютно верно. Основные причины пропусков - это технические проблемы и человеческий фактор. А можешь привести, пожалуйста?
</div>

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Потому что медиана представляет нам "типичное" значение в наборе данных. Также медиана в отличии от среднего устойчива к выборке, если в начале или в конце пристутсвуют аномальные значения (0.00001, 1, 2, 3, 4, 100). "Он ест мясо, Я капусту, вместе мы едим голубцы" медиана помогает избежать такого.

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера ❌<a class="tocSkip"></h3>

Выбор стоял между средним и медианой. В каких случаях выбор медианы предпочтительнее среднего? Поясни пожалуйста, поподробнее, какие характеристики выборки мотивируют нас использовать медиану
</div>

 <div class="alert alert-success">
<h3> Комментарий ревьюера v2 ✔️<a class="tocSkip"> </h3>

 Верно, такие выбивающиеся значения ещё называют выбросами.
</div>

### Шаг 4: общий вывод.

В ходе исследования надёжности заемщиков были изучены полученные данные, обработаны и заполнены пропуски, удалены дубликаты и клиенты были объединены в группы по разным параметрам. По итогам проделанной работы по поставленным вопросам можно сделать следующие выводы:

- **Наличие детей напрямую влияет на платеж по кредиту, риск семье с 3 и более детьми остаться должниками 9.3%.**
    
    * Возможно проблема с выборкой, так как репрезентативность группы "многодетные" очень мала, стоит пересмотреть данные.


- **Заемщики, не женаты/ не замужем или несостоящие в отношениях, имеют максимальный процент невозврата кредита(9.8% и 9.3%)соответственно, чем заемщики в браке.**
   
   * Репрезентативность выборки высокая, с данными проблем нет.

- **Основываясь на данных, которые мы смогли достать и категоризовать, то заемщики с доходом 50000-200000(С) имеют больше   надеждность для кредитования, чем другие.**
    
    * Репрезентативность доходов заемщиков очень низкая, возможно были ошибки при выгрузке данных, нужно перепроверить.
    
- **Категория граждан, берущая кредит на автомобиль или образование больше подвержены быть должниками(9.3% для обоих категорий).**
    * Тут стоит учесть еще возраст кредиторов, так как люди покупающие недвижимость или вступающие в брак, это не всегда люди с состоявшимся капиталом.

**Рекомендации:**
  Надежнее выдавать кредиты клиентам с отсутствием детей и семейным положением "женат / замужем", данные категории с больше вероятностью смогут погасить кредиты в установленные договором сроки. Так же такие категории клиентов чаще берут кредиты, выгода для банка на лицо.

<div class="alert alert-success">
<h3> Комментарий ревьюера v3 ✔️<a class="tocSkip"> </h3>

Теперь получился развёрнутый хороший вывод, учитывающий все показатели и этапы проекта. Отличная работа!
    
</div>

<div class="alert alert-block alert-danger">
<h3> Комментарий ревьюера v2 ❌<a class="tocSkip"></h3>

Хороший структурированный вывод. Но как я писал ранее, этот этап должен подробно описывать данные. Добавь, пожалуйста, значения, которые соответствуют описываемым группам, а также информацию о проблемах, найденных на этапе предобработки и пути их решения.
</div>

<div class="alert alert-danger">
    <h3> Комментарий ревьюера ❌<a class="tocSkip"> </h3>
    
Все этапы исследовательской части рассмотрены. Но не хватает деталей.
        
Стоит учитывать, что твой потенциальный руководитель, в первую очередь, обратит внимание на общий вывод. Поэтому важно отразить в нём все основные результаты из всех шагов проекта в числах. И важно описать проблемы, найденные на этапе предобработки, и пути их решения
        
Можно опираться на такую структуру финального вывода во всех проектах:
* предобработка данных с описанием найденных проблем и путей их решения
* ответы на цели проекта
* рекомендации для заказчика

Чтобы лучше понимать, как оформлять комментарии в markdown-ячейках, советую изучить [краткое руководство по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/).

</div>